File: PythonCodes/Chapter2/iteration.py
def for_loop(n: int) -> int:
    """for 循环"""
    res = 0
    # range(a, b) ==> a, a+1, ..., b-1
    for i in range(1, n + 1):
        res += i
    return res

def while_loop(n: int) -> int:
    """while 循环"""
    res = 0
    i = 1
    while i <= n:
        res += i
        i += 1
    return res

def while_loop_ii(n: int) -> int:
    """while 循环（两次更新）"""
    res = 0
    i = 1
    while i <= n:
        res += i
        i += 1
        i *= 2
    return res

def nested_for_loop(n: int) -> str:
    """双层循环"""
    res = ""
    for i in range(1, n + 1):
        for j in range(1, n + 1):
            res += f"({i}, {j})"
            if i != n or j != n:
                res += ", "
    return res

if __name__ == "__main__":
    n = 5
    res = nested_for_loop(n)
    print(f"\nfor 循环的求和结果 res = {res}")

File: CppCodes/Chapter2/iteration.cpp
#include <iostream>
#include <sstream>

/* for 循环 */
int forLoop(int n)
{
    int res = 0;
    for (int i = 1; i <= n; i++)
    {
        res += i;
    }
    return res;
}

/* while 循环 */
int whileLoop(int n)
{
    int res = 0;
    int i = 1;
    while (i <= n)
    {
        res += i;
        i++;
    }
    return res;
}

/* while 循环（两次更新）*/
int whileLoopII(int n)
{
    int res = 0;
    int i = 1;
    while (i <= n)
    {
        res += i;
        i += 1;
        i *= 2;
    }
    return res;
}

/* 双层 for 循环 */
std::string nestedForLoop(int n)
{
    std::string res = "";
    for (int i = 1; i <= n; i++)
    {
        for (int j = 1; j <= n; j++)
        {
            res += "(" + std::to_string(i) + ", " + std::to_string(j) + ")";
            if (i != n || j != n) // check if it is a last iteration
            {
                res += ", ";
            }
            // std::string str1 = "(" + std::to_string(i);
            // std::string str2 = ", ";
            // std::string str3 = std::to_string(j) + "), ";

            // res += (str1 + str2 + str3);
        }
    }
    return res;
}

/*
 * std::stringstream可以减少内存分配和拷贝的次数，因为它会在内部维护一个缓冲区
 * 而不是每次都分配新的内存空间进行字符串拼接。
 * 使用<<操作符可以直接向std::stringstream中添加数据，使得代码更加清晰易读。
 * std::stringstream可以方便地进行不同类型的数据转换，例如将整数、浮点数等转换为字符串。
 */
std::string nestedForLoopII(int n)
{
    std::stringstream res;
    for (int i = 1; i <= n; i++)
    {
        for (int j = 1; j <= n; j++)
        {
            res << "(" << i << ", " << j << ")";
            if (i != n || j != n) // check if it is a last iteration
            {
                res << ", ";
            }
        }
    }
    return res.str();
}

int main()
{
    // int res = 0;
    std::string res = "";
    res = nestedForLoopII(5);
    // printf("%s", res);
    printf("%s", res.c_str());
    // std::cout << res;
    return 0;
}